src/bellpepper/r1cs.rs

```diff
@@ -72,15 +71,14 @@ macro_rules! impl_nova_shape {
             &constraint.2,
           );
         }
-
         assert_eq!(num_cons_added, num_constraints);
 
-        let S: R1CSShape<G> = {
-          // Don't count One as an input for shape's purposes.
-          let res = R1CSShape::new(num_constraints, num_vars, num_inputs - 1, &A, &B, &C);
-          res.unwrap()
-        };
+        A.cols = num_vars + num_inputs;
```

Comment:
I'm a bit confused here. Shouldn't the number of columns be `num_vars` + `num_inputs` + 1? (In R1CS matrices, there is an extra column for the constant term 1).

Reply:
This is the dual of [@porcuquine 's comment](https://github.com/microsoft/Nova/pull/232/files#r1353508230) on passing `num_inputs - 1` to `R1CSShape` (as the code did): in the (unchanged) Nova codebase notation one is counted among `num_inputs` when it comes to the matrix.

This is checked e.g. in [`test_recursive_circuit_*`](https://github.com/microsoft/Nova/blob/9165211d352b6e5f9bcaef2e6cc5cb79666ecf6f/src/circuit.rs#L378), which calls [`is_sat`](https://github.com/microsoft/Nova/blob/9165211d352b6e5f9bcaef2e6cc5cb79666ecf6f/src/circuit.rs#L424) and hence performs a matrix-vector multiplication with the witness vector extended by 1:
https://github.com/microsoft/Nova/blob/9165211d352b6e5f9bcaef2e6cc5cb79666ecf6f/src/r1cs.rs#L250

This in turns calls for a check of the vector-matrix dimension match in `multiply_vec`:
https://github.com/microsoft/Nova/pull/232/files/86947fc8ea9fdfd8d9ebc31ca39fcb839dc67550#diff-0d99bd627c139dba595a777fc8b1b8da96465d7f1a03ff1b1010663cfa1d806dR78-R86
(note the difference with `multiply_vec_unchecked`).

Reply:
Thanks for clarifying!